scapy/contrib/oam.py

```diff
@@ -0,0 +1,673 @@
+        )
+
+
+class MegId(Packet):
+    """
+    MEG ID
+    """
+
+    name = "MEG ID"
+
+    fields_desc = [
+        ByteField("resv", 1),
+        ByteField("format", 0),
+        MultipleTypeField(
+            [
+                (
+                    LenField("length", 13, fmt="B"),
+                    lambda p: p.format == 32,
+                ),
+                (
+                    LenField("length", 15, fmt="B"),
+                    lambda p: p.format == 33,
+                )
+            ],
+            LenField("length", 45, fmt="B"),
+        ),
+        PadField(
+            MultipleTypeField(
+                [
+                    (
+                        FieldListField("values", [0] * 13,
+                                       ByteField("value", 0),
+                                       count_from=lambda pkt: pkt.length),
+                        lambda x: x.format == 32,
+                    ),
+                    (
+                        FieldListField("values", [0] * 15,
+                                       ByteField("value", 0),
+                                       count_from=lambda pkt: pkt.length),
+                        lambda x: x.format == 33,
+                    )
+                ],
+                NBytesField("values", 0, sz=45),
+            ),
+            45),
+    ]
+
+    def extract_padding(self, s):
+        return b"", s
+
+
+class OAM_TLV(Packet):
+    """
+    OAM TLV
+    """
+
+    name = "OAM TLV"
+    fields_desc = [ByteField("type", 1), LenField("length", None)]
+
+    def extract_padding(self, s):
+        return s[:self.length], s[self.length:]
+
+
+class OAM_DATA_TLV(Packet):
+    """
+    OAM Data TLV
+    """
+
+    name = "OAM Data TLV"
+    fields_desc = [ByteField("type", 3), LenField("length", None)]
+
+    def extract_padding(self, s):
+        return s[:self.length], s[self.length:]
+
+
+class OAM_TEST_TLV(Packet):
+    """
+    OAM test TLV data
+    """
+
+    name = "OAM test TLV"
+
+    fields_desc = [
+        ByteField("type", 32),
+        MultipleTypeField(
+            [
+                (
+                    LenField("length", None, adjust=lambda l: l + 5),
+                    lambda p: p.pat_type == 1 or p.pat_type == 3,
+                )
+            ],
+            LenField("length", None, adjust=lambda l: l + 1),
+        ),
+        EnumField(
+            "pat_type",
+            0,
+            {
+                0: "Null signal without CRC-32",
+                1: "Null signal with CRC-32",
+                2: "PRBS 2^-31 - 1 without CRC-32",
+                3: "PRBS 2^-31 - 1 with CRC-32",
+            },
+            fmt="B",
+        ),
+        ConditionalField(
+            FCSField("crc", None, fmt="!I"),
+            lambda p: p.pat_type == 1 or p.pat_type == 3,
+        ),
+    ]
+
+    def do_dissect(self, s):
+        if ord(s[3:4]) == 1 or ord(s[3:4]) == 3:
+            # move crc to the end of packet
```

Comment:
We actually have something fancy for this now, it's called `FCSField`, have you looked into it?

Reply:
Yes, and I'm using it. It's actually a hack to make it work because it takes data from the *end of the packet* and this layer may be a part of list.
In other words, I have this
```
||packet1|crc1||packet2|crc2||packet3|crc3||
```
and I have to transform it into this
```
||packet1||packet2||packet3||crc3||crc2||crc1||
```
to make `FCSField` work correctly.

Reply:
That's not super scapy-like, wouldn't a list of crcs work? Since you're using this inside a TLV, it would make sense to put all the checksums in a layer in the upper packet. You can then move the post_build logic there

Reply:
If this works it's fine.